File: cpp/1329_SortTheMatrixDiagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        for (int j = 0; j < mat[0].size(); ++j)
            diagonal_sort(mat, 0, j);
        
        for (int i = 0; i < mat.size(); ++i)
            diagonal_sort(mat, i, 0);
        
        return mat;
    }
    
    void diagonal_sort(vector<vector<int>>& m, int row, int col) {
        vector<int> entries;
        int i = row, j = col;
        while (i < m.size() && j < m[0].size())
            entries.push_back(m[i++][j++]);
        
        sort(entries.begin(), entries.end());
        i = row, j = col;
        for (int e : entries)
            m[i++][j++] = e;
    }
};

File: python/1329_SortTheMatrixDiagonally.py
class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        def sort_diagonal(row, col):
            nonlocal n, m
            i, j = row, col
            diagonal = []
            while i < m and j < n:
                heappush(diagonal, mat[i][j])
                i += 1
                j += 1
            
            while diagonal:
                mat[row][col] = heappop(diagonal)
                row += 1
                col += 1
        
        m, n = len(mat), len(mat[0])
        for i in range(m):
            sort_diagonal(i, 0)
        
        for j in range(n):
            sort_diagonal(0, j)
            
        return mat
